### ISSUE TITLE
False negative: `ExplicitItLambdaParameter`

### ISSUE THREAD

**Issue by SomeTroglodyte** [detekt/detekt#6822]
## Expected Behavior
Typing a lambda `it` to allow inference is sometimes the simplest way to enable shorthands like [SAM conversion](https://kotlinlang.org/docs/fun-interfaces.html).
```kotlin
        val dir = com.badlogic.gdx.files.FileHandle(".")
        val subdirs = dir.list { it: File -> it.isDirectory }
```
...is calling [this list() function](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/files/FileHandle.java#L441), successfully, and succinctly. Without typing the `it`, inference fails.

A rewrite that doesn't trip detekt would definitely be far less readable:
```kotlin
        val dir = com.badlogic.gdx.files.FileHandle(".")
        @Suppress("ObjectLiteralToLambda")
        val subdirs = dir.list(object : FileFilter { override fun accept(file: File) = file.isDirectory })
```

## Observed Behavior
Analysis fails with `This explicit usage of `it` as the lambda parameter name can be omitted. [ExplicitItLambdaParameter]`

## Steps to Reproduce
see above

## Context
I encountered a slightly more complex example earlier in https://github.com/detekt/detekt/pull/6367#issuecomment-1687561697 - and rewrote that source.
Job showing this one: https://github.com/yairm210/Unciv/actions/runs/7426175374/job/20209368340?pr=10881
Issue search turns up a few, but none comparable, most fixed - but this is not, as 1.23.4 already contains PR #6367, right?

## Your Environment

* Version of detekt used: 1.23.4
* Github workflow definition: https://github.com/yairm210/Unciv/blob/cfff0aa66a2bb05d3da002bad54695bae4951d40/.github/workflows/detektAnalysis.yml
* Link to project fork (if it's a public repository): https://github.com/SomeTroglodyte/Unciv

**Comment by atulgpt** [comment]
Hi, @SomeTroglodyte why can't you use the below code? In my case removing `it` and directly using `val subdirs2 = dir.list { it.isDirectory }` working fine to me.

```
fun interface FileFilter {
    fun accept(file: File): Boolean
}

class FileHandle {
    fun list(filter: FileFilter): Array<FileHandle?> {
        return emptyArray()
    }
}

fun main() {
    val dir = FileHandle()
    val subdirs1 = dir.list { it: File -> it.isDirectory } // this is problematic code
    val subdirs2 = dir.list { it.isDirectory } // this is working fine
}
```

**Comment by BraisGabin** [comment]
This have sense. I would vote to allow any `it` parameter if it declares the type.

**Comment by BraisGabin** [comment]
From the rule description:

> [...] Naming it just `it` is meaningless and only makes your code misleading, especially when dealing with nested functions.

meaningless:

```kotlin
{ it -> /* something */ }
```

No meaningless:
```kotlin
{ it: File -> /* something */ }
```

The type gives information that could help read the code. It is not meaningless. I agree that `it` is probably not a good name, but this rule is not about that.
